clients/drcachesim/common/trace_entry.h

```diff
@@ -634,6 +634,15 @@ type_is_data(const trace_type_t type)
         type == TRACE_TYPE_DATA_FLUSH_END;
 }
 
+/**
+ * Returns whether the type represents a memory read access.
+ */
+static inline bool
+type_is_read(const trace_type_t type)
+{
+    return type_is_prefetch(type) || type == TRACE_TYPE_READ;
```

Comment:
Hmm, what about cache flushes which are part of instr_is_data() and instr_has_address()?  If you leave them out of here, what if someone uses instr_is_data() or instr_has_address() and then assumes !type_is_read() means it's a store and so assumes all the cache flushes are stores?  If left out, probably the doc comment should explicitly say it excludes them, to avoid confusion vs type_has_address().

UPDATE: I see your own code does this!

As for whether we should check cache flushes: certainly x86 OP_clflush has a memory operand (marked as a source, so a load/read) we can check.  There are unresolved issues with the a64 decoder though IIRC: #4400 has discussion and linked issues/PRs.  It might be messy to try to check them if the deoder is inconsistent -- but OTOH if the instrumentation is using the same operands to get the address for the record it should work in the checker too.

Reply:
I've added cache flushes to type_is_read.